presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java

```diff
@@ -540,13 +532,7 @@ private String getVarcharValue(Type type, Object value)
             if (type instanceof VarcharType) {
                 return ((Slice) value).toStringUtf8();
             }
-            if (type instanceof TinyintType || type instanceof SmallintType || type instanceof IntegerType || type instanceof BigintType) {
-                return ((Long) value).toString();
-            }
-            if (type instanceof BooleanType) {
-                return ((Boolean) value).toString();
-            }
-            throw new PrestoException(NOT_SUPPORTED, format("Unsupported data type in EXPLAIN (TYPE IO): %s", type.getDisplayName()));
+            return Objects.toString(value);
```

Comment:
@bhhari Does this work for all types? For example, is this correct for arrays, maps, structs, geometry types, etc.?

How is the output of EXPLAIN (TYPE IO) is used? Is there any implicit assumption that only partition columns are specified in the output? Will this change break that assumption?

Reply:
@mbasmanova  this works for all the complex types arrays, structs, maps with pushdown,
before this change they are not part of the current constraint and they are not retrieved.
The type IO is to see which table we are reading from with column constraints and which table we are outputing to.

Reply:
@bhhari Hmm... what would this return for complex types? I see that value can be a Slice or a Block in which case toString() won't produce anything meaningful. The other question is whether the system that is receiving the output of EXPLAIN (TYPE IO) is ready to handle the output where non-partition columns have filters. Would you double check that it doesn't have this assumption baked in and will accept the new output?

Reply:
I agree that we should first check the return value of complex type before supporting it. 

Since `pushdownFilterEnabled` is default false, I think it should be fine because the default behavior is not changed. Of course it would be nice if we could only keep partition column predicate as output for `EXPLAIN IO` even if we enable filter pushdown, but I'm not sure how hard that would be as I'm not familiar with filter push down.

Reply:
here is how it looks for complex types
ARRAY
```
                Query Plan
------------------------------------------
 {
   "inputTableColumnInfos" : [ {
     "table" : {
       "catalog" : "hive",
       "schemaTable" : {
         "schema" : "tpch",
         "table" : "lineitem_ex"
       }
     },
     "columnConstraints" : [ {
       "columnName" : "keys",
       "typeSignature" : "array(bigint)",
       "domain" : {
         "nullsAllowed" : true,
         "ranges" : [ ]
       }
     } ]
   } ]
 }

```

STRUCT
 ```
                                                             Query Plan
----------------------------------------------------------------------------------------------------------------------------------------
 {
   "inputTableColumnInfos" : [ {
     "table" : {
       "catalog" : "hive",
       "schemaTable" : {
         "schema" : "tpch",
         "table" : "lineitem_ex"
       }
     },
     "columnConstraints" : [ {
       "columnName" : "info",
       "typeSignature" : "row(orderkey bigint,linenumber integer,shipdate row(ship_day tinyint,ship_month tinyint,ship_year integer))",
       "domain" : {
         "nullsAllowed" : true,
         "ranges" : [ ]
       }
     } ]
   } ]
 }

```

Map

```
                   Query Plan
------------------------------------------------
 {
   "inputTableColumnInfos" : [ {
     "table" : {
       "catalog" : "hive",
       "schemaTable" : {
         "schema" : "tpch",
         "table" : "test_maps"
       }
     },
     "columnConstraints" : [ {
       "columnName" : "map_keys",
       "typeSignature" : "map(integer,bigint)",
       "domain" : {
         "nullsAllowed" : false,
         "ranges" : [ ]
       }
     } ]
   } ]
 }

```

Reply:
> I agree that we should first check the return value of complex type before supporting it.
> 
> Since `pushdownFilterEnabled` is default false, I think it should be fine because the default behavior is not changed. Of course it would be nice if we could only keep partition column predicate as output for `EXPLAIN IO` even if we enable filter pushdown, but I'm not sure how hard that would be as I'm not familiar with filter push down (maybe we need to keep original constraint and new constraint? I'm not sure if this is a good idea at all...).

How is this information used ? Like is there any optimization done based on the column contraints, by not providing all the constraints we are already giving limited info ( is the original behavior right?) .  What is impacted if we add other constraints?

Reply:
Currently we directly send the constraints to an internal routing system so they could route the query to the data center where we would have best data locality (there are other considerations as well but IO info is used for locality). I don't have a deep understanding of how the routing system works internally.

I expect the impact to be small, especially we are using a separate tier of nodes to process explain queries which feed the routing system, and filter pushdown is not enabled there.

I cannot say for external usage outside of the scope of this internal routing system.